### ISSUE TITLE
4.2-General Guide-Model Bias, Optimization Issue, and Overfitting, Validation 

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#28]
當你在訓練模型時, 判斷現在問題的方法
![image](https://user-images.githubusercontent.com/34474924/235036674-461216b2-e7b3-46b4-bc16-ba9f51765aea.png)

首先最重要的一定是要判斷training data上的Loss, 所以一定要記錄training loss

# Training Loss太大可能的原因Model Bias, Optimization Issue

_______________________________________________

## Model Bias

model太簡單→大海撈針但針不在海裡
![image](https://user-images.githubusercontent.com/34474924/235040272-0145314c-8423-4522-a984-0120c055ba02.png)

解決：設一個更大的model，增加model的彈性，例如增加feature, piecewise linear curve, 增加Neuron
![image](https://user-images.githubusercontent.com/34474924/235040320-c28c9a9c-9766-4ef8-922f-c12bc4507152.png)

______________

## Optimization Issue

![image](https://user-images.githubusercontent.com/34474924/235043351-bf034ec9-04de-4f08-a481-59eab42d0012.png)

目前此門課只會使用Gradient Descent，而找不到Local minimum→大海撈針，撈不到

![image](https://user-images.githubusercontent.com/34474924/235043569-1d9baa81-1bbd-4001-a8cd-87ff8eef9a83.png)

_________________

## Model Bias v.s. Optimization Issue

Gaining the insights from comparison 
![image](https://user-images.githubusercontent.com/34474924/235044115-143a690f-da58-4f94-a661-c28a17cbc0f3.png)

56層 v.s. 20層 的測試資料，20層的loss<56層的loss, 單看測試資料可能是overfitting, 但並不一定是, 要確認training data
如果training error 56層<20層, 但是 test error 56層>20層, 這樣就有可能是overfitting
因為在訓練的時候，越深/複雜的model應該要越契合訓練資料，甚至會overfitting，所以訓練資料的loss，深<淺

然而56層 v.s. 20層 的訓練資料，20層的loss<56層的loss, 由此可知不是overfitting
此時也不可能是Model Bias 因為56層的彈性比20層高

所以看training error就可知, 會造成這種現象一定是 `Optimization Issue`

先用簡單的model來做訓練，再做深的model，如果發現training loss，深>淺→optimiaztion Issue

**Comment by Pin-Jiun** [comment]
# Overfitting

![image](https://user-images.githubusercontent.com/34474924/235046059-70e24eda-de16-4c0d-b64f-6481a812c199.png)

綠色線就是Overfitting造成的結果, 選用黑色的線才是較好的選擇
```
overfitting iff Small loss on training data, large loss on testing data
```
解決
1. 增加訓練資料

- 尋找新的資料(然而這是最難的步驟, 效率也很差先不考慮)
- Data augmentation→對現有資料做更動，變成(假)新的資料
![image](https://user-images.githubusercontent.com/34474924/235046714-92944d1f-fc31-4ed0-afaf-67f09bcae287.png)


2. 給model一些限制，最好是跟題目要求有關的限制。

- Less parameters, sharing parameters 
- Less features (Feature 不是越多越好)
- Early stopping  
- Regularization
- Dropout 


e.g.已知題目可能會是2次函數，就把model限制在2次式
然而給模型太大的限制→可能會回到Model Bias的問題

Bias-Complexity Trade-off
![image](https://user-images.githubusercontent.com/34474924/235047063-be3d43cd-f4de-4fe6-8c73-b05ba1c21e47.png)

_________

HW中testing set會分成 public和private
![image](https://user-images.githubusercontent.com/34474924/235052610-2020295a-dbcf-4b93-aa78-8adb405149e7.png)

不該追求public testing error小, 而過度training model
因為你訓練出來的model有可能是符合(擬合)public test , 但是和實際的(private testing data)狀況有差
這樣儘管你的training error小且testing error小也會造成現實狀況(未知, 未來的測試)造成嚴重的錯誤

這樣到底要怎麼選擇好的model參數呢?

________

# Validation Set 驗證集

應該使用驗證集取代觀看公開測試集的方式
![image](https://user-images.githubusercontent.com/34474924/235060725-f694e461-c899-464a-97b3-9eb3d5983fe7.png)


正確流程
```
1. 將training set分成 training set和Validation Set
2. 使用training set訓練model, 經過Validation Set計算Loss
3. 當Validation Set計算的Loss小時才更新參數(表示較好的model)並儲存, 避免overfitting
4. 最後看testing error相對較低的地方選擇儲存的model, 避免overfitting
```

可以說Validation Set是雙重的保證, 避免overfitting testing set和training set

但如果隨機分, 有可能造成model很好, 但是validation set的結果很差怎辦?
使用 N-fold Cross Validation

## N-fold Cross Validation

以3-fold為例
分三份，每份都分兩份訓練&一份驗證，與model交叉測試
![image](https://user-images.githubusercontent.com/34474924/235061060-32e28176-114b-43a5-bb31-05645ce246d1.png)

```
Split your training data into training set and validation set for model selection

```
____________

# Mismatch
當訓練資料跟測試資料不一樣時，訓練資料的品質就不重要了。
Your training and testing data have different distributions. 

e.g.訓練資料是彩色真實動物圖片，測試資料是黑白卡通動物圖騰。

HW 11才會出現
![image](https://user-images.githubusercontent.com/34474924/235061626-f2526eaa-cc34-4316-bf6f-a4b5e164f111.png)
